plugins/variable-import-plugins/complex-variable-flattener/README.md

```diff
@@ -0,0 +1,214 @@
+// imports come here
+
+public class ComplexVariableFlattener implements VariableImportAdapter {
+  
+  /*
+   some fields
+  */
+
+  @Override
+  public List<PluginVariableDto> adaptVariables(List<PluginVariableDto> variables) {
+    List<PluginVariableDto> resultList = new ArrayList<>();
+    for (PluginVariableDto pluginVariableDto : variables) {
+      logger.debug("adapting variable {} (v{}) of process-instance {}...", //
+                   pluginVariableDto.getName(), //
+                   pluginVariableDto.getVersion(), //
+                   pluginVariableDto.getProcessInstanceId());
+      if (pluginVariableDto.getType().equalsIgnoreCase("object")) {
+        String serializationDataFormat = String.valueOf(pluginVariableDto.getValueInfo().get("serializationDataFormat"));
+        if (serializationDataFormat.equals("application/json")) {
+          this.flatJsonObject(pluginVariableDto, resultList);
+        } else {
+          logger.warn("complex variable '{}' won't be imported (unsupported serializationDataFormat: {})",
+                      pluginVariableDto.getName(), serializationDataFormat);
+        }
+      } else {
+        resultList.add(pluginVariableDto);
+      }
+    }
+    return resultList;
+  }
+
+  private void flatJsonObject(PluginVariableDto variable, List<PluginVariableDto> resultList) {
+    if (variable.getValue() == null || variable.getValue().isEmpty()) {
+      return;
+    }
+
+    try {
+      new JsonFlattener(variable.getValue()) //
+        .withFlattenMode(FlattenMode.KEEP_ARRAYS) //
+        .flattenAsMap() //
+        .entrySet() //
+        .stream() //
+        .map(e -> this.map(e.getKey(), e.getValue(), variable)) //
+        .filter(Optional::isPresent) //
+        .map(Optional::get) //
+        .forEach(resultList::add);
+    } catch (Throwable t) {
+      logger.error("error while flattening variable '" + variable.getName() + "')", t);
+    }
+  }
+  
+  /*
+   additional mapping helper methods
+  */
+
+}
+```
+
+This code would flatten any JSON object passed to it by converting each field to a dot-separated primitive variable.
+So the following user object:
+  ```
+  {
+   "firstName": "John",
+   "lastName": "Doe"
+  }
+  ```
+will get transformed to two primitive variables of the type String: `user.firstName: "John"` & `user.lastName: "Doe"`.
+
+### Build the plugin and add it to your Optimize distribution
+
+Now run the following command to build the Jar containing the implemented plugin:
+
+```cmd
+mvn clean install
+```
+
+This creates a `complex-variable-flattener-X.X.X.jar` in `target` folder. Copy this
+jar file into the `plugin` folder of your Optimize distribution.
+
+Finally it is necessary to tell Optimize in which package it should search for the plugin. You
+could do that if you add the following lines to `environment-config.yaml` in the
+`config` folder of your Optimize distribution:
+
+```yaml
+plugin:
+  variableImport:
+    basePackages: ['de.janhuddel.bpm.optimize.plugin']
```

Comment:
this would need to change too if we do above. If it's fine to keep as is, then feel free to ignore both comments